src/irr/video/COpenGLSpecializedShader.cpp

```diff
@@ -114,10 +114,104 @@ COpenGLSpecializedShader::COpenGLSpecializedShader(const video::IVideoDriver* _d
 
     GLchar logbuf[1u<<12]; //4k
     driver->extGlGetProgramInfoLog(m_GLname, sizeof(logbuf), nullptr, logbuf);
-    os::Printer::log(logbuf, ELL_ERROR);
+    if (logbuf[0])
+        os::Printer::log(logbuf, ELL_ERROR);
 
-    // TODO:
-    // what should be interface for setting push_constants (regular uniform on GL backend) for now?
+    m_introspectionData = core::smart_refctd_ptr<asset::CIntrospectionData>(_introspection);
+}
+
+void COpenGLSpecializedShader::setUniformsImitatingPushConstants(const uint8_t* _pcData)
+{
+    if (m_uniformsList.empty())
+        buildUniformsList();
+
+    using gl = COpenGLExtensionHandler;
+    for (const SUniform& u : m_uniformsList)
+    {
+        const SMember& m = u.m;
+        auto is_mtx = [&m] { return (m.mtxRowCnt>1u && m.mtxColCnt>1u); };
+        auto is_single_or_vec = [&m] { return (m.mtxRowCnt>=1u && m.mtxColCnt==1u); };
+
+        if (is_mtx() && m.type==asset::EGVT_F32) {
+            core::vector<GLfloat> matrix_data(m.mtxRowCnt*m.mtxColCnt*m.count);
```

Comment:
`vector` is a malloc, use an `std::array` with an upper bound in size

Reply:
or some stack scratch space

Reply:
I would if someone could give me upper bound :P really, is there any expectable?

Reply:
128 bytes?

Reply:
so max size of 4x4 matrix array is 2? i dont get it

Reply:
yes, if our max size of push constant data supported by the engine is 128, then you can have max 2 matrices.

Reply:
i was thinking totally in abstraction of this.. Ok, doable if we have some global constant with max push constants bytesize

Reply:
yeah best that way, so no alloc here, this is the hot loop

Reply:
resolved